File: backend/device_actions.py
from typing import Dict, Any, Optional, List
import re
import logging

logger = logging.getLogger(__name__)

class DeviceActionParser:
    """Parse user messages to detect device control intents"""
    
    @staticmethod
    def parse_message(message: str) -> Dict[str, Any]:
        """
        Parse user message to detect device control commands
        Returns action type and parameters
        """
        message_lower = message.lower()
        
        # SMS/Text message detection
        if any(keyword in message_lower for keyword in ['text', 'message', 'sms', 'send']):
            # Extract phone number
            phone_match = re.search(r'\d{10,}', message)
            phone_number = phone_match.group(0) if phone_match else None
            
            # Extract contact name
            name_match = re.search(r'(?:text|message|sms)\s+([A-Z][a-z]+(?:\s+[A-Z][a-z]+)?)', message)
            contact_name = name_match.group(1) if name_match else None
            
            # Extract message content
            message_content = DeviceActionParser._extract_message_content(message)
            
            if phone_number or contact_name:
                return {
                    'action': 'sms',
                    'phone_number': phone_number,
                    'contact_name': contact_name,
                    'message': message_content,
                    'needs_confirmation': True,
                }
        
        # Call detection
        if 'call' in message_lower and 'called' not in message_lower:
            phone_match = re.search(r'\d{10,}', message)
            name_match = re.search(r'call\s+([A-Z][a-z]+(?:\s+[A-Z][a-z]+)?)', message)
            
            if phone_match or name_match:
                return {
                    'action': 'call',
                    'phone_number': phone_match.group(0) if phone_match else None,
                    'contact_name': name_match.group(1) if name_match else None,
                    'needs_confirmation': True,
                }
        
        # Camera
        if any(keyword in message_lower for keyword in ['camera', 'photo', 'picture', 'take a pic']):
            return {
                'action': 'camera',
                'needs_confirmation': False,
            }
        
        # Contacts
        if 'contact' in message_lower or 'phone number' in message_lower:
            name_match = re.search(r'(?:for|of|find)\s+([A-Z][a-z]+(?:\s+[A-Z][a-z]+)?)', message)
            return {
                'action': 'contacts',
                'contact_name': name_match.group(1) if name_match else None,
                'needs_confirmation': False,
            }
        
        # Music
        if any(keyword in message_lower for keyword in ['play music', 'play song', 'open spotify', 'open youtube music']):
            song_match = re.search(r'play\s+(.+?)(?:\s+on|\s+from|$)', message_lower)
            return {
                'action': 'music',
                'query': song_match.group(1) if song_match else None,
                'needs_confirmation': False,
            }
        
        # No device action detected
        return {'action': 'none'}
    
    @staticmethod
    def _extract_message_content(text: str) -> Optional[str]:
        """Extract message content from quotes or after keywords"""
        patterns = [
            r'["\'](.+?)["\']',  # Content in quotes
            r'(?:saying|say|tell them|tell him|tell her)\s+(.+)',  # After keywords
        ]
        
        for pattern in patterns:
            match = re.search(pattern, text, re.IGNORECASE)
            if match:
                return match.group(1)
        
        return None
    
    @staticmethod
    def generate_confirmation_message(action_data: Dict[str, Any], personality: str) -> str:
        """
        Generate AI response that includes device action confirmation
        """
        action = action_data.get('action')
        
        if personality == 'tharos':
            if action == 'sms':
                contact = action_data.get('contact_name') or action_data.get('phone_number')
                message = action_data.get('message', '')
                return f"Yo, got it! I'll help you text {contact}. Want me to send: '{message}'? Just say yes and I'll open it for you."
            
            elif action == 'call':
                contact = action_data.get('contact_name') or action_data.get('phone_number')
                return f"Alright bro, calling {contact} now. Hold on..."
            
            elif action == 'camera':
                return "Opening the camera for you. Smile!"
            
            elif action == 'contacts':
                return "Let me check your contacts for you..."
            
            elif action == 'music':
                query = action_data.get('query', 'music')
                return f"Yo, playing {query} for you now!"
        
        else:  # superintendent
            if action == 'sms':
                contact = action_data.get('contact_name') or action_data.get('phone_number')
                message = action_data.get('message', '')
                return f"Understood. I will compose a message to {contact}: '{message}'. Shall I proceed?"
            
            elif action == 'call':
                contact = action_data.get('contact_name') or action_data.get('phone_number')
                return f"Initiating call to {contact} now, sir."
            
            elif action == 'camera':
                return "Opening the camera application for you."
            
            elif action == 'contacts':
                return "Accessing your contacts directory..."
            
            elif action == 'music':
                query = action_data.get('query', 'your music')
                return f"Playing {query} for you now."
        
        return ""

File: frontend/services/deviceActions.ts
import * as SMS from 'expo-sms';
import * as Contacts from 'expo-contacts';
import * as Camera from 'expo-camera';
import * as MediaLibrary from 'expo-media-library';
import { Linking, Alert, Platform } from 'react-native';

// Device action types
export type DeviceAction = {
  type: 'sms' | 'call' | 'contacts' | 'camera' | 'music' | 'open_app' | 'none';
  params?: any;
};

export type DeviceActionResult = {
  success: boolean;
  message: string;
  error?: string;
};

/**
 * Parse AI response to detect device control commands
 */
export function parseDeviceAction(aiResponse: string, userMessage: string): DeviceAction {
  const lowerMessage = userMessage.toLowerCase();
  const lowerResponse = aiResponse.toLowerCase();
  
  // SMS/Text message detection
  if (lowerMessage.includes('text') || lowerMessage.includes('message') || lowerMessage.includes('sms')) {
    // Extract phone number and message
    const phoneMatch = userMessage.match(/\d{10,}/);
    if (phoneMatch) {
      return {
        type: 'sms',
        params: {
          phoneNumber: phoneMatch[0],
          message: extractMessageContent(userMessage),
        },
      };
    }
    
    // Check if AI response suggests sending a message
    if (lowerResponse.includes('send') || lowerResponse.includes('text')) {
      return {
        type: 'sms',
        params: {},
      };
    }
  }
  
  // Call detection
  if (lowerMessage.includes('call') && !lowerMessage.includes('called')) {
    const phoneMatch = userMessage.match(/\d{10,}/);
    if (phoneMatch) {
      return {
        type: 'call',
        params: { phoneNumber: phoneMatch[0] },
      };
    }
  }
  
  // Contact search
  if (lowerMessage.includes('contact') || lowerMessage.includes('phone number')) {
    return {
      type: 'contacts',
      params: {},
    };
  }
  
  return { type: 'none' };
}

function extractMessageContent(text: string): string {
  // Try to extract message content after keywords
  const patterns = [
    /(?:text|message|sms).*?"([^"]+)"/i,
    /(?:text|message|sms).*?'([^']+)'/i,
    /(?:say|tell).*?"([^"]+)"/i,
  ];
  
  for (const pattern of patterns) {
    const match = text.match(pattern);
    if (match) return match[1];
  }
  
  return '';
}

/**
 * Execute device actions
 */
export async function executeDeviceAction(action: DeviceAction): Promise<DeviceActionResult> {
  try {
    switch (action.type) {
      case 'sms':
        return await sendSMS(action.params);
      
      case 'call':
        return await makeCall(action.params);
      
      case 'contacts':
        return await searchContacts(action.params);
      
      case 'camera':
        return await openCamera(action.params);
      
      case 'music':
        return await openMusic(action.params);
      
      case 'open_app':
        return await openApp(action.params?.url || action.params);
      
      default:
        return {
          success: false,
          message: 'Unknown action type',
        };
    }
  } catch (error: any) {
    return {
      success: false,
      message: 'Action failed',
      error: error.message,
    };
  }
}

/**
 * Send SMS
 */
async function sendSMS(params: { phoneNumber?: string; message?: string }): Promise<DeviceActionResult> {
  try {
    // Check if SMS is available
    const isAvailable = await SMS.isAvailableAsync();
    if (!isAvailable) {
      return {
        success: false,
        message: 'SMS is not available on this device',
      };
    }
    
    const { phoneNumber, message } = params;
    
    if (!phoneNumber) {
      return {
        success: false,
        message: 'Phone number is required',
      };
    }
    
    // Open SMS composer
    const result = await SMS.sendSMSAsync(
      [phoneNumber],
      message || ''
    );
    
    return {
      success: true,
      message: 'SMS composer opened',
    };
  } catch (error: any) {
    return {
      success: false,
      message: 'Failed to open SMS composer',
      error: error.message,
    };
  }
}

/**
 * Make phone call
 */
async function makeCall(params: { phoneNumber: string }): Promise<DeviceActionResult> {
  try {
    const { phoneNumber } = params;
    
    if (!phoneNumber) {
      return {
        success: false,
        message: 'Phone number is required',
      };
    }
    
    const url = `tel:${phoneNumber}`;
    const canOpen = await Linking.canOpenURL(url);
    
    if (!canOpen) {
      return {
        success: false,
        message: 'Cannot make calls on this device',
      };
    }
    
    await Linking.openURL(url);
    
    return {
      success: true,
      message: `Calling ${phoneNumber}`,
    };
  } catch (error: any) {
    return {
      success: false,
      message: 'Failed to initiate call',
      error: error.message,
    };
  }
}

/**
 * Search contacts
 */
async function searchContacts(params: { query?: string }): Promise<DeviceActionResult> {
  try {
    // Request permission
    const { status } = await Contacts.requestPermissionsAsync();
    
    if (status !== 'granted') {
      return {
        success: false,
        message: 'Contacts permission denied',
      };
    }
    
    // Get all contacts
    const { data } = await Contacts.getContactsAsync({
      fields: [Contacts.Fields.PhoneNumbers, Contacts.Fields.Name],
    });
    
    if (data.length > 0) {
      const contactList = data.slice(0, 5).map((contact, index) => {
        const phone = contact.phoneNumbers?.[0]?.number || 'No phone';
        return `${index + 1}. ${contact.name}: ${phone}`;
      }).join('\n');
      
      return {
        success: true,
        message: `Found ${data.length} contacts. First 5:\n${contactList}`,
      };
    }
    
    return {
      success: true,
      message: 'No contacts found',
    };
  } catch (error: any) {
    return {
      success: false,
      message: 'Failed to access contacts',
      error: error.message,
    };
  }
}

/**
 * Open camera
 */
async function openCamera(params: any): Promise<DeviceActionResult> {
  try {
    // Request camera permission
    const { status } = await Camera.Camera.requestCameraPermissionsAsync();
    
    if (status !== 'granted') {
      return {
        success: false,
        message: 'Camera permission denied. Please enable camera access in settings.',
      };
    }
    
    // Note: Opening camera requires navigation to a camera screen
    // For now, we'll just confirm permission is granted
    return {
      success: true,
      message: 'Camera permission granted. Camera feature ready!',
    };
  } catch (error: any) {
    return {
      success: false,
      message: 'Failed to access camera',
      error: error.message,
    };
  }
}

/**
 * Open music apps
 */
async function openMusic(params: { query?: string; app?: string }): Promise<DeviceActionResult> {
  try {
    const { query, app } = params || {};
    
    let url = '';
    
    // Determine which music app to open
    if (app && app.toLowerCase().includes('spotify')) {
      url = query ? `spotify:search:${encodeURIComponent(query)}` : 'spotify:';
    } else if (app && app.toLowerCase().includes('youtube')) {
      url = query 
        ? `vnd.youtube://results?search_query=${encodeURIComponent(query)}`
        : 'vnd.youtube://';
    } else {
      // Default: Try to search on YouTube Music or regular YouTube
      url = query 
        ? `https://music.youtube.com/search?q=${encodeURIComponent(query)}`
        : 'https://music.youtube.com';
    }
    
    const canOpen = await Linking.canOpenURL(url);
    
    if (!canOpen) {
      // Fallback to web URL
      const webUrl = query 
        ? `https://www.youtube.com/results?search_query=${encodeURIComponent(query)}`
        : 'https://www.youtube.com';
      
      await Linking.openURL(webUrl);
      return {
        success: true,
        message: query ? `Searching for "${query}" on YouTube` : 'Opening YouTube',
      };
    }
    
    await Linking.openURL(url);
    
    return {
      success: true,
      message: query ? `Playing "${query}"` : 'Opening music app',
    };
  } catch (error: any) {
    return {
      success: false,
      message: 'Failed to open music app',
      error: error.message,
    };
  }
}

/**
 * Open app by URL scheme (Android/iOS)
 */
export async function openApp(appScheme: string): Promise<DeviceActionResult> {
  try {
    const canOpen = await Linking.canOpenURL(appScheme);
    
    if (!canOpen) {
      return {
        success: false,
        message: `Cannot open ${appScheme}`,
      };
    }
    
    await Linking.openURL(appScheme);
    
    return {
      success: true,
      message: `Opened ${appScheme}`,
    };
  } catch (error: any) {
    return {
      success: false,
      message: 'Failed to open app',
      error: error.message,
    };
  }
}